tests/core/snapd-refresh-vs-services/task.yaml

```diff
@@ -0,0 +1,212 @@
+    snap install --dangerous snapd_2.49.2.snap
+    snap version | MATCH 2.49.2
+
+  elif [ "${SNAPD_VERSION_UNDER_TEST}" = "pr" ]; then
+      # refresh back to the version we originally had from before the test 
+      # started
+      echo "Refreshing snapd to version from the pr"
+      snap install --dangerous snapd-pr.snap
+  fi
+
+  # now install another service that will either get Requires= or Wants= for
+  # usr-lib-snapd.mount, depending on the variant of the test
+  "$TESTSTOOLS"/snaps-state install-local-as test-snapd-simple-service test-snapd-simple-service_alt
+
+  # check that it is still initially active
+  snap services|MATCH ".*test-snapd-simple-service_alt.test-snapd-simple-service\s*enabled\s*active.*"
+
+  # if we are not running with 2.49.2, then the second service should have been
+  # generated with Wants=, but if we are doing 2.49.2 then the second service 
+  # should have Requires= in it
+  if [ "${SNAPD_VERSION_UNDER_TEST}" = "2.49.2" ];then 
+    MATCH Requires=usr-lib-snapd.mount < /etc/systemd/system/snap.test-snapd-simple-service_alt.test-snapd-simple-service.service
+  else
+    MATCH Wants=usr-lib-snapd.mount < /etc/systemd/system/snap.test-snapd-simple-service_alt.test-snapd-simple-service.service
+  fi
+
+  # now repack current snapd to refresh to it again and observe the results
+  unsquashfs -d ./snapd-unpacked snapd-pr.snap
+  snap pack --filename=snapd_repacked.snap snapd-unpacked
+
+  # save the PID's of the services
+  systemctl show -p MainPID snap.test-snapd-simple-service.test-snapd-simple-service > old-main.pid
+  systemctl show -p MainPID snap.test-snapd-simple-service_alt.test-snapd-simple-service > old-main_alt.pid
+
+  echo "Refresh snapd"
+  snap install --dangerous snapd_repacked.snap
+
+  # for all variants of the test, the main.pid should be the same since it did
+  # not ever contain the Requires=
+  systemctl show -p MainPID snap.test-snapd-simple-service.test-snapd-simple-service > new-main.pid
+
+  if [ "$(cat new-main.pid)" != "$(cat old-main.pid)" ]; then
+    echo "The service without Requires= was restarted, test is broken"
+    exit 1
+  fi
+
+  # for the variants of the test that have the fix in them, that is stable and
+  # and pr, then we also should not have had the alt service pid change due to a
+  # restart
+  # for the other variant (2.49.2 only), we unfortunately ended up needing to
+  # restart the service so it should have a different PID, but it should be 
+  # running again
+
+  systemctl show -p MainPID snap.test-snapd-simple-service_alt.test-snapd-simple-service > new-main_alt.pid
+  if [ "${SNAPD_VERSION_UNDER_TEST}" = "2.49.2" ]; then
+    if [ "$(cat new-main_alt.pid)" = "$(cat old-main_alt.pid)" ]; then
+      echo "Somehow the service was not killed as expected ... test is probably broken"
+      exit 1
+    fi
+  else
+    if [ "$(cat new-main.pid)" != "$(cat old-main.pid)" ]; then
+      echo "The service with Wants= was unexpectedly killed test is broken"
```

Comment:
nitpick: better add a `;` or a `-` before "test is broken". Though I actually wonder if we need this "test is broken" sentence: the test will fail anyway, and this line does not add a lot of info.

Reply:
sure I can add some punctuation, as the old saying goes punctuation saves lives: "let's eat, grandma" vs "let's eat grandma"

I do prefer to keep it in, because this is a specific failure case that is important to call out specifically, and including the phrasing "test is broken" hopefully makes it more obvious to the reader that the test did not fail due to flakiness somewhere else or some transient issue and that if it fails like this it really should be investigated and not just ignored and restarted

Reply:
done